app/models/form.rb

```diff
@@ -154,48 +154,80 @@ def invalidated_pages(log, current_user = nil)
+  def reset_not_routed_checkbox_questions(log)
+    checkbox_questions = routed_and_not_routed_questions_by_type(log, type: "checkbox")
+    checkbox_questions[:invalid].each do |invalidated_question|
+      valid_options = checkbox_questions[:valid]
+                                        .select { |q| q.id == invalidated_question.id }
+                                        .flat_map { |q| q.answer_options.keys }
+      invalidated_question.answer_options.each_key do |invalid_option|
+        if !log.respond_to?(invalid_option) || valid_options.include?(invalid_option)
+          next
         else
-
-          (question.answer_options.keys - enabled_answer_options.keys).map do |invalid_answer_option|
-            Rails.logger.debug("Cleared #{invalid_answer_option} value")
-            log.public_send("#{invalid_answer_option}=", nil) if log.respond_to?(invalid_answer_option)
-          end
+          clear_attribute(log, invalid_option)
         end
+      end
+    end
+  end
+
+  def reset_radio_questions_if_not_routed_or_invalid_answers(log)
+    radio_questions = routed_and_not_routed_questions_by_type(log, type: "radio")
+    valid_radio_options = radio_questions[:valid]
+                                   .group_by(&:id)
+                                   .transform_values! { |q_array| q_array.flat_map { |q| q.answer_options.keys } }
+    radio_questions[:invalid].each do |invalidated_question|
+      q_id = invalidated_question.id
+      if !log.respond_to?(q_id) || log.public_send(q_id).nil? || valid_radio_options.key?(q_id)
+        next
       else
-        Rails.logger.debug("Cleared #{question.id} value")
-        log.public_send("#{question.id}=", nil) unless enabled_question_ids.include?(question.id)
+        clear_attribute(log, q_id)
+      end
+    end
+    valid_radio_options.each do |q_id, valid_options|
+      if !log.respond_to?(q_id) || valid_options.include?(log.public_send(q_id).to_s)
+        next
+      else
+        clear_attribute(log, q_id)
+      end
+    end
+  end
+
+  def reset_not_routed_free_user_input_questions(log)
+    non_radio_checkbox_questions = routed_and_not_routed_questions_by_type(log)
+    enabled_question_ids = non_radio_checkbox_questions[:valid].map(&:id)
+    non_radio_checkbox_questions[:invalid].each do |invalidated_question|
+      q_id = invalidated_question.id
+      if log.public_send(q_id).nil? || enabled_question_ids.include?(q_id)
+        next
+      else
+        clear_attribute(log, q_id)
       end
     end
   end
 
-  def enabled_page_questions(log)
-    questions - invalidated_page_questions(log)
+  def routed_and_not_routed_questions_by_type(log, type: nil, current_user: nil)
+    # we're already treating these fields as a special case and reset their values upon saving a log
+    callback_questions = %w[postcode_known la ppcodenk previous_la_known prevloc postcode_full ppostcode_full location_id address_line1 address_line2 town_or_city county]
```

Comment:
I support this and I think that this list is different categories of things and it might be better to interrogate what they are and why.
eg. they are mostly radio questions where the options are populated on render I believe. For those questions the best way to get that information form the Form that occurs to me is questions that have no answer options when this code runs. Do you think that is robust?
then I will look into the other questions , but for instance I do not know why postcode_known is in this list, I assume that at some point for some other ticket someone didn't want to it to be cleared and just put it here without thinking further.